### DISCUSSION TITLE
Patterns for loading user-supplied jars

### DISCUSSION THREAD

**Discussion by mark-burch** [hydraulic-software/conveyor#125] [Q&A] [answered]
First of all, huge thank you for Conveyor. We're trying to convert our current jpackage arrangement to Conveyor and it's evident how well thought out every aspect is. 

We have an app where external jars (e.g JDBC drivers) need to be loaded into the JVM. We'd prefer these jars sit outside of the installation directory and be under the user's Application Support in our directory where we store some other files on MacOS. 

Our current jpackage setup has us adding that custom directory into the class path and leverages the fact that there's an APP_DIR variable that expands into the installation directory so we can travel up from that. This doesn't work on MacOS as applications get copied to the absolute /Applications folder where we can't find the user's home.

Wondering if there is some magic in Conveyor that would allow us to add a dynamic directory (AppData in Windows, ~/Applications Support/... in MacOS) to the class path or whether a different approach would be more warranted (perhaps URLClassloading from within the app?).

Thanks in advance!

**Comment by mikehearn** [comment]
Thank you for the kind words, we appreciate it!

The usual way you'd implement this is using classloaders. The simplest approach is to simply instantiate the Driver class directly using the normal classloader mechanisms. Unfortunately JDBC driver _discovery_ is somewhat classloader hostile and so the direct approach does assume you know the driver classname.

Being able to add home directory paths to the classpath is an interesting feature request and shouldn't be too hard to add. It does break code signing unfortunately, whereas with classloaders you can use signed JARs and check at load time. But this is a somewhat theoretical concern in today's code signing landscape.

**Comment by poonam-meghnani** [comment]
Hi, 
I'd like to upvote Mark's request for the feature enabling to add home or data directory to the class-path. I am working with Mark on the problem he mentioned above and using class loader to load each driver does not feel like a permanent solution as it needs to be aware of driver classname. 
As Mark mentioned we are using a variable to refer to installation path in our current set up, but thats a limitation moving to conveyor as we cant get a guaranteed install-path for all installers. 
Could you please help us with feature or workaround that could be configured through conveyor? 
Thanks a ton!

**Reply by mikehearn** [reply]
So for this you'd like something like:

```
app.jvm.extra-classpath = [ "~/AppData/Local/MyApp/Plugins" ]
```

and for the launcher to resolve `~` to the user's home directory at startup?

**Reply by poonam-meghnani** [reply]
Thank you for the prompt response!

Thats right. we might not need launcher to resolve ~ as we could instead use ${env.HOME}?
The idea is to be able to add additional path to the class-path so that external libs could be resolved at startup. 
Also, I want to understand if this would be an OS specific config for the paths could be different?

**Reply by mikehearn** [reply]
I guess we'd have to make it platform-specific in this case, yes.

**Reply by poonam-meghnani** [reply]
I'd appreciate any information you could share about the timeline for this change please.

**Reply by mikehearn** [reply]
It will be in the next release, which will probably come out in a few weeks.

**Reply by poonam-meghnani** [reply]
Thank you, looking forward to it.

**Reply by mikehearn** [reply]
Hi,

Could you please try this beta?

https://downloads.hydraulic.dev/beta/conveyor/download.html

It understands how to expand ~/ or ~\ at the start of an element of the classpath. I didn't go for full environment variable interpolation in the end for various reasons.

**Reply by poonam-meghnani** [reply]
> Hi,
> 
> Could you please try this beta?
> 
> https://downloads.hydraulic.dev/beta/conveyor/download.html
> 
> It understands how to expand ~/ or ~\ at the start of an element of the classpath. I didn't go for full environment variable interpolation in the end for various reasons.

Thank you Mike, 
I appreciate your help on this!

Could you please help me with an example of adding a dynamic directory to the class-path?
for example: if I want to add ~/AppData/Local/MyApp/Drivers to app's class-path so that the driver classes on this path loads up  when the app is started.
I am wondering if I could simply set it as mentioned in chat above:
`app.jvm.extra-classpath = [ "~/AppData/Local/MyApp/Drivers" ]`

**Reply by mikehearn** [reply]
The key you want is in the launcher object:

https://conveyor.hydraulic.dev/17.0/configs/jvm/#launchers

So it'd be something like:

```
app.jvm {
    gui {
        main-class = com.example.MyMainClass
        options = [ ... jvm flags go here ... ]
        class-path = [ "*.jar", "~/AppData/Local/MyApp/Drivers", "~/Library/Application Data/"${app.rdns-name}"/Drivers" ]
    } 
}
```

Or however you want to do it. You can just add the paths for each platform to every platform for simplicity. Make sure not to forget the default "*.jar" setting to continue picking up the jars shipped with your app.

**Reply by poonam-meghnani** [reply]
Sure, thank you! 
I will try it and confirm soon.

**Reply by poonam-meghnani** [reply]
Hey Mike,
I tried the beta version for adding dynamic directory to class-path. I added the below in the sample config:
jvm {
    gui {
      main-class = "com.mongodb.migrator.application.MyApp"
      class-path = [ "*.jar", "~/Library/Application Support/MyApp/Drivers"]
....
}
}

- When the path ends at dir-name "~/Library/Application Support/MyApp/Drivers", it does not load the required classes from the dir.

- When the path has match string like "~/Library/Application Support/MyApp/Drivers/*.jar", it does not load the required classes from the dir.

- When I add the file name - "~/Library/Application Support/MyApp/Drivers/mysql-connector-j-8.3.0.jar", it works! 

However, I might not know the exact library name as user could be running a different version, I cannot hard-code the name. IS there a way to add everything (*.jar) on the class-path?

Thank you!

**Reply by mikehearn** [reply]
Ah, right. It's just expanding the tilde and the rest is up to the JVM, which doesn't like expanding globs. The first entry works because it's expanded by Conveyor at package build time. OK, will get this fixed for the next beta. Should be a quick fix.

**Reply by poonam-meghnani** [reply]
Thank you!

**Reply by mikehearn** [reply]
Please try 17.100.5 which should expand `*.jar` at the end of a classpath entry.

**Reply by poonam-meghnani** [reply]
Hey Mike, 
I am not able to run the conveyor 17.100.5 the way I did the older versions. It broke after updating to the newer version, below is the error on executing simple make command for macOS:
```
Exception in thread "main" java.lang.ClassNotFoundException: hydraulic.conveyor.app.ConveyorApplication
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:467)
	at hydraulic.conveyor.launcher/hydraulic.conveyor.launcher.Launcher.prepare(Launcher.java:57)
```
I get same error even when I ran - `conveyor --help`. I also cleared the cache but no success.
When I revert it back to 17.100.2, it works.

**Reply by mikehearn** [reply]
Yes that beta was DOA, sorry, please try 17.100.5

**Reply by poonam-meghnani** [reply, marked as answer]
> Yes that beta was DOA, sorry, please try 17.100.5

Hello Mike,
I tried 17.100.6 and it works perfect! Though, I have tried it only for MacOS aarch64 installer, I trust it should not be any different for other platforms.
Thank you for working through it.

**Reply by mikehearn** [reply]
Great. We should be able to get the current release out as stable next week or a bit later. I hope that isn't too much delay for you.

**Reply by poonam-meghnani** [reply]
Sounds good, thank you!

**Reply by mikehearn** [reply]
Conveyor 18 has been released with this feature included.
